docs: add STYLE_GUIDE.md for rules doc comments (closes #150) (#364)
falcosecurity/rules: docs/STYLE_GUIDE.md

Reviewer: Shouldn't this be inside docs/ for mkdocs to find it? 🤔

with the substance.

### GOOD

```yaml
# Detects process name existence; required before identity-based conditions.
# Handles the edge case of dropped syscall events where proc.name may be "<NA>".
# TODO: Remove the "N/A" variant once scap-file compatibility is no longer needed.
- macro: proc_name_exists
  condition: (not proc.name in ("<NA>","N/A"))
```

- First sentence states the purpose directly.
- Second sentence explains the rationale.
- TODO is explicit and actionable.

### BAD

```yaml
# This macro is about the process name
- macro: proc_name_exists
  condition: (not proc.name in ("<NA>","N/A"))
```

"About the process name" tells a reviewer nothing they couldn't infer from
the macro name itself.

---

## 3. Tuning Macros (`user_*`, `allowed_*`)

Tuning macros are placeholders — their default condition is `(never_true)` or
`(always_true)`, and operators are expected to override them locally to suppress
false positives. Because these macros have no `desc` field, their doc comment is
the *only* documentation an operator will read before deciding how to override.

**Required in every tuning macro doc comment:**

1. What the macro does (one sentence).

> 2. How to override it — include a concrete `override: append` snippet.

Reviewer 2: ```suggestion
2. How to override it — include a concrete `override:` block snippet (e.g. `condition: append`).
```

Same syntax point as the inline comments below 🙏

3. The field or fields most commonly used in overrides (e.g., `proc.name`,
   `container.image.repository`).

### GOOD

```yaml
# Suppresses "Read sensitive file untrusted" for known-legitimate readers.
# Default is never_true (no suppression). To allow specific processes:
#
#   - macro: user_known_read_sensitive_files_activities
#     condition: (proc.name in (my_backup_agent, my_audit_tool))

> #     override: append
> #
> # Use `override: replace` to swap the default condition entirely instead of
> # extending it. Common override fields: proc.name, proc.exepath,
> # container.image.repository.

Reviewer 2: ```suggestion
#     override:
#       condition: append
#
# Use `condition: replace` (instead of `append`) to swap the default condition
# entirely. Common override fields: proc.name, proc.exepath,
# container.image.repository.
```

`override` is a block in the canonical syntax (the value selects per-field append/replace). For a macro it is `condition: append`; for a list, `items: append`. See 👉 https://falco.org/docs/concepts/rules/overriding/

I verified with `falco -V` on 0.43.0: the single-value form (`override: append`) loads but produces schema warnings, so I would rather we teach the right shape from day one 🙏

- macro: user_known_read_sensitive_files_activities
  condition: (never_true)
```

### BAD

```yaml
# Add conditions here to suppress false positives.
- macro: user_known_read_sensitive_files_activities
  condition: (never_true)
```

An operator hitting a false positive at 2 AM needs the override snippet, not
a paraphrase of what overriding means.

---

### GOOD — `allowed_*` style (negative-logic placeholder)

```yaml
# Allowlist of process names permitted to initiate outbound SSH connections.
# Default is never_true; the rule uses double negation
# (and not allowed_ssh_initiators), so this macro effectively allows nothing
# when unset and the rule's negative gate stays open. Operators override to
# narrow the set of expected SSH-initiating processes.
#
# To restrict SSH initiation to known automation tooling:
#
#   - macro: allowed_ssh_initiators
#     condition: (proc.name in (ansible-playbook, scp, rsync))

> #     override: append
> #
> # Use `override: replace` to swap the default condition entirely instead of
> # extending it. Common override fields: proc.name, proc.exepath, fd.rip
> # (for peer-IP allowlists). See also: macro never_true for the placeholder
> # pattern.

Reviewer 2: ```suggestion
#     override:
#       condition: append
#
# Use `condition: replace` (instead of `append`) to swap the default condition
# entirely. Common override fields: proc.name, proc.exepath, fd.rip
# (for peer-IP allowlists). See also: macro never_true for the placeholder
# pattern.
```

Same here 🙏

# The open_write/open_read macros cover the relevant cases via the open* family.
# Re-enable if a future Falco version surfaces raw read/write as distinct events.
#
# - macro: write
#   condition: (syscall.type=write and fd.type in (file, directory))
# - macro: read
#   condition: (syscall.type=read and fd.type in (file, directory))
```

### BAD — unexplained comment

```yaml
# - macro: write
#   condition: (syscall.type=write and fd.type in (file, directory))
# - macro: read
#   condition: (syscall.type=read and fd.type in (file, directory))
```

A reviewer seeing this has no idea if the macros are intentionally disabled,
temporarily scaffolded, or simply left from a refactor.

---

### GOOD — tuning template with opt-in explanation

```yaml
# Optional: treat any node process in a container as a protected shell spawner.
# Disabled by default because node is also widely used as a build-pipeline tool
# where spawning shells is expected. Enable by overriding this macro to remove
# the never_true guard:
#
#   - macro: possibly_node_in_container
#     condition: (proc.pname=node and proc.aname[3]=docker-containe)

> #     override: replace   # swap the default never_true guard entirely

Reviewer 2: ```suggestion
#     override:
#       condition: replace   # swap the default never_true guard entirely
```

Same here 🙏

```

### BAD — section header attached to first item

```yaml
# Shell-spawning protection macros and lists
- list: protected_shell_spawning_binaries
  items: [...]
```

---

## 7. Field Order within a Rule

The [Falco Rules Style Guide](https://falco.org/docs/rules/style-guide/) is
authoritative on field ordering. For reference, the order used in existing
`maturity_stable` rules is:

```yaml
- rule: <name>
  desc: >
    <description>
  condition: >
    <condition>
  output: <output string>
  priority: <PRIORITY>
  tags: [<maturity_tag>, <workload_tags>, <mitre_tags>]
  enabled: <true|false>   # omit when true (the default)
  source: <source>        # omit when syscall (the default)
```

`enabled: false` and `source:` are placed after `tags` and only included when
they differ from the default. This ordering is a convention observed in the
existing rule corpus; defer to maintainers if guidance changes.

---

## Quick Reference

| Situation | Convention |
|-----------|------------|
| Comment immediately precedes an item (no blank line) | Doc comment — documents that item |
| Comment has a blank line before it AND below it | Section comment — labels a group |

> | Tuning macro (`user_*`, `allowed_*`) | Doc comment must include override snippet with `override: append` (or `override: replace`) |

Reviewer 2: ```suggestion
| Tuning macro (`user_*`, `allowed_*`) | Doc comment must include override snippet using an `override:` block (e.g. `condition: append` or `condition: replace`) |
```

Same syntax fix as above 🙏

| List with generated or sourced content | Doc comment must include source command or URL |
| Commented-out item with no explanation | Delete it — use git history |
| Commented-out item kept intentionally | Doc comment must state why and when to re-enable |
| First sentence of any doc comment | Summarize purpose; avoid "This macro…" preamble |

---

## Relationship to the Existing Style Guide

This guide covers **comment conventions only**. For rule expression style,
output format, tag requirements, and maturity-level criteria, see the
[Falco Rules Style Guide](https://falco.org/docs/rules/style-guide/) and

> [CONTRIBUTING.md](CONTRIBUTING.md).

Reviewer 2: ```suggestion
[CONTRIBUTING.md](../CONTRIBUTING.md).
```

`CONTRIBUTING.md` is at the repo root, so from here it needs `../`. Otherwise the relative link breaks 🙏
